crates/ra_proc_macro_srv/src/dylib.rs

```diff
@@ -0,0 +1,205 @@
+//! Handles dynamic library loading for proc macro
+
+use crate::{proc_macro::bridge, rustc_server::TokenStream};
+use std::path::Path;
+
+use goblin::{mach::Mach, Object};
+use libloading::Library;
+use ra_proc_macro::ProcMacroKind;
+
+static NEW_REGISTRAR_SYMBOL: &str = "_rustc_proc_macro_decls_";
+
+fn get_symbols_from_lib(file: &Path) -> Option<Vec<String>> {
+    let buffer = std::fs::read(file).ok()?;
+    let object = Object::parse(&buffer).ok()?;
+
+    return match object {
+        Object::Elf(elf) => {
+            let symbols = elf.dynstrtab.to_vec().ok()?;
+            let names = symbols.iter().map(|s| s.to_string()).collect();
+            Some(names)
+        }
+        Object::PE(pe) => {
+            let symbol_names =
+                pe.exports.iter().flat_map(|s| s.name).map(|n| n.to_string()).collect();
+            Some(symbol_names)
+        }
+        Object::Mach(mach) => match mach {
+            Mach::Binary(binary) => {
+                let exports = binary.exports().ok()?;
+                let names = exports
+                    .iter()
+                    .map(|s| {
+                        // In macos doc:
+                        // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dlsym.3.html
+                        // Unlike other dyld API's, the symbol name passed to dlsym() must NOT be
+                        // prepended with an underscore.
+                        if s.name.starts_with("_") {
+                            s.name[1..].to_string()
+                        } else {
+                            s.name.to_string()
+                        }
+                    })
+                    .collect();
+                Some(names)
+            }
+            Mach::Fat(_) => None,
+        },
+        Object::Archive(_) | Object::Unknown(_) => None,
+    };
+}
+
+fn is_derive_registrar_symbol(symbol: &str) -> bool {
+    symbol.contains(NEW_REGISTRAR_SYMBOL)
+}
+
+fn find_registrar_symbol(file: &Path) -> Option<String> {
+    let symbols = get_symbols_from_lib(file)?;
+
+    symbols.iter().find(|s| is_derive_registrar_symbol(s)).map(|s| s.clone())
```

Comment:
```suggestion
    symbols.into_iter().find(|s| is_derive_registrar_symbol(s))
```

Untested. Sorry for the nits.

Reply:
Fixed. thanks.